lib/route/ROUTING.md

```diff
@@ -0,0 +1,446 @@
+# Obelisk Routing
+
+The `obelisk-routes` package is designed to help with managing the paths and parameters for
+routing in your application.
+
+This is distinct from packages like `servant` that are designed for specifying endpoints for a
+REST API. When building routes with `obelisk-routes` you're not going to be handling headers or
+HTTP methods.
+
+One of the goals of `obelisk-routes` is to allow for complex routing structures to be partitioned
```

Comment:
While this is certainly a goal, most of the complexity of obelisk-routes isn't related to this goal.  The big thing we are trying to do here is provide several important guarantees:

- *Every* route value can be encoded to a URL (i.e. encoding can't fail)
- `decode . encode = pure` (i.e.: encoding roundtrips, which also implies that there's no ambiguity in the routes)
- When you make common modifications to routes, such as adding or removing a route, you get "incomplete case" and similar warnings everywhere you need to update your app - this makes it much harder to miss things when you add routes.

When I was initially investigating the options for route parsing, I couldn't find anything that achieved all three of these, and most options (e.g. boomerang) didn't achieve *any* of them.

One thing to keep in mind here is that this library asks a lot from the user, e.g. not allowing arbitrary shadowing.  This means we need to give them a lot of value in return.  Personally, I consider these guarantees to be really valuable, and I think we should probably make sure they're mentioned prominently, especially when they help explain limitations or difficult stuff.

Reply:
Intro has been rewritten as a 'motivation' section that contains these points. :+1: It may need to be fleshed out a bit with some descriptive examples ?